// aws-cdk-lib/aws-ec2/lib/machine-image/amazon-linux-2023.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonLinux2023ImageSsmParameter = exports.AmazonLinux2023Kernel = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("./common");
/**
 * Amazon Linux 2023 kernel versions
 */
class AmazonLinux2023Kernel {
    constructor(version) {
        this.version = version;
    }
    /**
     * Generate a string representation of the kernel
     */
    toString() {
        return `kernel-${this.version}`;
    }
}
_a = JSII_RTTI_SYMBOL_1;
AmazonLinux2023Kernel[_a] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2023Kernel", version: "2.77.0" };
/**
 * The latest kernel version currently available in a published AMI.
 *
 * When a new kernel version is available for an al2023 AMI this will be
 * updated to contain the latest kernel version and will cause your instances
 * to be replaced. Do not store stateful information on the instance if you are
 * using this version.
 */
AmazonLinux2023Kernel.CDK_LATEST = new AmazonLinux2023Kernel('6.1');
/**
 * The default kernel version for Amazon Linux 2023 is 6.1 and
 * the SSM parameter does not include it in the name
 * (i.e. /aws/service/ami-amazon-linux-latest/amzn2023-ami-kernel-default-x86_64)
 */
AmazonLinux2023Kernel.DEFAULT = new AmazonLinux2023Kernel('default');
/**
 * Kernel version 6.1
 */
AmazonLinux2023Kernel.KERNEL_6_1 = new AmazonLinux2023Kernel('6.1');
exports.AmazonLinux2023Kernel = AmazonLinux2023Kernel;
/**
 * A SSM Parameter that contains the AMI ID for Amazon Linux 2023
 */
class AmazonLinux2023ImageSsmParameter extends common_1.AmazonLinuxImageSsmParameterBase {
    /**
     * Generates a SSM Parameter name for a specific amazon linux 2022 AMI
     *
     * Example values:
     *
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-6.1-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-6.1-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-minimal-kernel-6.1-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-minimal-kernel-6.1-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-minimal-kernel-default-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-minimal-kernel-default-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-arm64",
     */
    static ssmParameterName(props) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2023ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.ssmParameterName);
            }
            throw error;
        }
        const edition = (props && props.edition) || common_1.AmazonLinuxEdition.STANDARD;
        const parts = [
            common_1.AmazonLinuxGeneration.AMAZON_LINUX_2023,
            'ami',
            edition !== common_1.AmazonLinuxEdition.STANDARD ? edition : undefined,
            (props.kernel ?? AmazonLinux2023Kernel.DEFAULT)?.toString(),
            props.cpuType ?? common_1.AmazonLinuxCpuType.X86_64,
        ].filter(x => !!x);
        return '/aws/service/ami-amazon-linux-latest/' + parts.join('-');
    }
    constructor(props = {}) {
        super({
            parameterName: AmazonLinux2023ImageSsmParameter.ssmParameterName(props),
            cachedInContext: props.cachedInContext,
            userData: props.userData,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2023ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, AmazonLinux2023ImageSsmParameter);
            }
            throw error;
        }
    }
}
_b = JSII_RTTI_SYMBOL_1;
AmazonLinux2023ImageSsmParameter[_b] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2023ImageSsmParameter", version: "2.77.0" };
exports.AmazonLinux2023ImageSsmParameter = AmazonLinux2023ImageSsmParameter;
